planner/core/task.go

```diff
@@ -161,6 +161,85 @@ func (p *PhysicalApply) attach2Task(tasks ...task) task {
 	}
 }
 
+func (p *PhysicalIndexMergeJoin) attach2Task(tasks ...task) task {
+	innerTask := p.innerTask
+	outerTask := finishCopTask(p.ctx, tasks[p.OuterIndex].copy())
+	if p.OuterIndex == 0 {
+		p.SetChildren(outerTask.plan(), innerTask.plan())
+	} else {
+		p.SetChildren(innerTask.plan(), outerTask.plan())
+	}
+	p.schema = buildPhysicalJoinSchema(p.JoinType, p)
+	return &rootTask{
+		p:   p,
+		cst: p.GetCost(outerTask, innerTask),
+	}
+}
+
+// GetCost computes the cost of index merge join operator and its children.
+func (p *PhysicalIndexMergeJoin) GetCost(outerTask, innerTask task) float64 {
+	var cpuCost float64
+	outerCnt, innerCnt := outerTask.count(), innerTask.count()
+	// Add the cost of evaluating outer filter, since inner filter of index join
+	// is always empty, we can simply tell whether outer filter is empty using the
+	// summed length of left/right conditions.
+	if len(p.LeftConditions)+len(p.RightConditions) > 0 {
+		cpuCost += cpuFactor * outerCnt
+		outerCnt *= selectionFactor
+	}
+	// Cost of extracting lookup keys.
+	innerCPUCost := cpuFactor * outerCnt
+	// Cost of sorting and removing duplicate lookup keys:
+	// (outerCnt / batchSize) * (sortFactor + 1.0) * batchSize * cpuFactor
+	// If `p.NeedOuterSort` is false, the sortFactor is batchSize * Log2(batchSize).
+	// Otherwise, it's 0.
+	batchSize := math.Min(float64(p.ctx.GetSessionVars().IndexJoinBatchSize), outerCnt)
+	sortFactor := 0.0
+	if p.NeedOuterSort {
+		sortFactor = math.Log2(float64(batchSize))
+	}
+	if batchSize > 2 {
+		innerCPUCost += outerCnt * (sortFactor + 1.0) * cpuFactor
```

Comment:
If `p.NeedOuterSort` is true, when will the outer sort be done? is it in parallel? it decides whether we should amortize this part of cost to inner workers or not.

Reply:
It same as `IndexJoin`, because we need to construct the ranges of the inner reader with the order of outer rows.